File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {

  static boolean[] visited;
  static int answer;

  public int solution(int k, int[][] dungeons) {
    answer = 0;
    visited = new boolean[dungeons.length];

    return dfs(dungeons, k, visited, 0);
  }

  static int dfs(int[][] dungeons, int k, boolean[] visited, int cnt) {

    for (int i = 0; i < dungeons.length; i++) {
      if (!visited[i] && dungeons[i][0] <= k) {
        visited[i] = true;
        dfs(dungeons, k - dungeons[i][1], visited, cnt + 1);
        visited[i] = false;
      }
    }

    answer = Math.max(answer, cnt);

    return answer;
  }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    global answer
    answer = -1 
    visited = [False] * len(dungeons)
    dfs(dungeons, visited, k , 0)
    return answer


def dfs(dungeons, visited, k, cnt) :
    global answer 
    for i in range(len(dungeons)) :
        if not visited[i] and dungeons[i][0] <= k :
            visited[i] = True
            dfs(dungeons, visited, k - dungeons[i][1], cnt + 1)
            visited[i] = False
    
    answer = max(answer,cnt)
    